File: Lua/utils/primality/fermatPrimalityTest.lua
-- fermatPrimalityTest.lua
-- 
-- 

local mathUtil = require("utils.mathUtil")
local tests = { 2, 3 }


local function test(number)
	if number < 2 then return false end
	if number ~= 2 and number % 2 == 0 then return false end 
	
	for _, test in pairs(tests) do
		if mathUtil.modPow(test, number - 1, number) ~= 1 then
			return false
		end
	end
	
	return true
end

return test

File: CSharp/RandomNumberGeneration/BlumBlumShub/Utils/Primality/FermatPrimalityTest.cs
using System.Linq;

namespace RandomNumberGenerators.Utils.Primality {
    public class FermatPrimalityTest : IPrimalityTest {
        private static readonly long[] tests = new long[] { 2, 3 };

        public bool Test(long number) {
            if(number < 2) return false;
            if(number != 2 && number % 2 == 0) return false;

            foreach(var test in tests.Where(x => x < number)) {
                if(MathUtil.ModPow(test, number - 1, number) != 1) {
                    return false;
                }
            }

            return true;
        }
    }
}
